src/rgw/rgw_auth_s3.h

```diff
@@ -358,6 +361,14 @@ static inline std::string get_v4_canonical_uri(const req_info& info) {
 
   std::string canonical_uri = info.request_uri_aws4;
 
+  const size_t key_pos = canonical_uri.find("/", 1);
+  if (key_pos != std::string::npos) {
+    string bucket, key;
+    bucket = canonical_uri.substr(1, key_pos - 1);
+    key = canonical_uri.substr(key_pos + 1);
+    canonical_uri = std::string("/") + rgw::auth::s3::aws4_uri_encode(bucket) + std::string("/") + key;
```

Comment:
isn't `request_uri_aws4` assumed to be url-encoded already?

Reply:
rgw doesn't url encode it, but we assume the client sent it that way (`RGWREST::preprocess()` decodes it with `s->decoded_uri = url_decode(s->info.request_uri);`). this change would cause us to double-encode things that came in properly url-encoded. it sounds like the bug here is in the client instead?

Reply:
in boto https://github.com/boto/boto/blob/develop/boto/auth.py#L427 it encode the canonical_uri